### DISCUSSION TITLE
OpenDAL S3 only with 50Mbit/s Upload

### DISCUSSION THREAD

**Discussion by bebehei** [rustic-rs/rustic#1284] [Things are not working as expected - is this a bug?] [answered]
I'm currently testing S3 remote repositories and I'm comparing restic with rustic.

I'm doing tests with the initial backup of about 10GB and different compression schemes. After inital backup, I also do a test restore.

Now after checking my data, rustic seems to be exactly half as fast as restic.

When checking my prometheus data, it becomes clear:

![image](https://github.com/user-attachments/assets/c4faf1be-cc1b-4f92-ae32-17bf076ddfe3)

- I ran 3 tests with restic (from 15:50 to 16:35)
- I ran 4 tests with rustic (from 16:45 to 18:20)
- The dataset for restic and rustic is the same. The destination is also the same.
- As you can see, I've got 500Mbit/s down and 100Mbit/s up.
- restic uses the full 100Mbit/s for the inital backup, but rustic uses only about 50Mbit/s
- no memory/CPU pressure

Is there anything I can configure for OpenDAL, which might have limited it that way? I have searched the [S3 service documentation](https://opendal.apache.org/docs/rust/opendal/services/struct.S3.html), but I cannot find anything for limit/max connections.

Is there something else, I'm overlooking?

**Comment by simonsan** [comment]
Hey, thanks for opening that topic. To understand it right, you are comparing restic+rclone vs rustic w/opendal? Would you be able to do another measurement with rustic+rclone to understand on which side the issue might be?

**Reply by bebehei** [reply]
I'm using [native S3 from restic](https://restic.readthedocs.io/en/stable/030_preparing_a_new_repo.html#s3-compatible-storage), so I'm using the repo URL of `s3:https://fsn1.your-objectstorage.com/<bucket_name>`.

For rustic, I'm using OpenDAL as you have assumed.

I'll look forward to have an rclone test with rustic.

**Reply by bebehei** [reply]
I have done the tests with rclone, I'm stuck with 50Mbit/s.

I've tested the OpenDAL parameters `connections = "10"` (and varying other values) and `throttle = "1GB,1GB"` without success.

**Reply by simonsan** [reply]
> I have done the tests with rclone, I'm stuck with 50Mbit/s.

This implies rustic+rclone, right? Because it seems it is an issue with `rustic` then if either `rclone` and `opendal` don't change anything.

**Reply by bebehei** [reply]
I might miss rclone settings (AFAIK there is a connection parameter or something else).

Also during backup, there is every few seconds a short CPU spike from rustic doing read+compression+encryption and then going to sleep again. It's visible in htop and also in prometheus (if you set the scrape interval to ridiculous 1s).

The limit is clearly in the arbitrary network threshold.

Are there any connection parameters I can tune in rustic itself?

**Reply by bebehei** [reply]
Oh, I've overlooked your answer.


> This implies rustic+rclone, right? Because it seems it is an issue with `rustic` then if either `rclone` and `opendal` don't change anything.

Yes, this is right.

For me, it seems to be in the communication from rustic -> storage abstraction.

In NixOS 24.05, there is still version 0.7.0 included. I'm going to rule this out, too. I'm currently building v0.9.1 inside the provided docker container.

**Reply by bebehei** [reply, marked as answer]
Ok, I had been successfully now with a docker container and v0.9.1, yields full speed for about 100Mbit/s.

After building the docker container again with v0.7.0, I got the old result of 50Mbit/s. Seems like a real version problem there.

**Reply by simonsan** [reply]
Thanks for investigating this, though!
